### PR TITLE
Adds a `remove-corrupted-data` CLI tool command to recover a corrupted shard

### LINKED ISSUE #16561: Add CLI tool to recover a corrupted shard
### Problem Statement

In some rare cases, e.g. due to disk failures, some files of a shard can get corrupted. If there isn't a good copy of that shard somewhere else inside the cluster, CrateDB isn't able to recover from such failures.
The only solution to this situation is to remove the corrupted files so CrateDB will start the shard again. Of course this will result in partial data loss.

### Possible Solutions

Elasticsearch already ships a CLI tool to solve such situations, see https://www.elastic.co/guide/en/elasticsearch/reference/current/shard-tool.html (https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/shard/ShardToolCli.java).
We could port it to our code base (or rewrite it) to provide users a simple way to recover from this issue.

### Considered Alternatives

Remove the corrupted files manually from the file system.

Related discussion https://community.cratedb.com/t/translogcorruptedexception/1264

### PR DESCRIPTION
**Pull request by seut**
This tool will removed any corrupted parts of the Lucene index and translog of a shard, mostly resulting in a data loss. Only useful as the last resort on disaster recovery when no good copy (replica) exists or cannot be restored.

Closes #16561.

### CODE REVIEW

**Comment by seut** [comment]
The issues mentioned at https://github.com/crate/crate/pull/18217#issuecomment-3136535259 seems to be invalid, there is no need that shards must contain a `corruption` marker for the tool to run successfully. See also the pushed `RemoveCorruptedShardDataCommandIT` tests.

**Comment by seut** [comment]
I've manually tested that the new `crate-shard` script works as expected.

File: server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommand.java [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+            Arrays.asList(partitionedByValues),
+            true,
+            im -> im
+        ).getFirst();
+        if (indexMetadata == null) {
+            String partitionValString = partitionValues == null ? "" :
+                partitionValues.stream()
+                    .map(kvp -> kvp.key + "=" + kvp.value)
+                    .collect(Collectors.joining(","));
+            throw new ElasticsearchException("Unable to find table [" + indexName + "/" + partitionValString + "] in cluster state");
```
Comment (jeeminso) [inline server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommand.java]: We can use `getIndex(..)` variant that returns IndexMetadata.

File: server/src/main/java/io/crate/analyze/AlterTableRerouteAnalyzer.java [hunk 2]
```diff
@@ -140,6 +141,29 @@ public RerouteAnalyzedStatement visitRerouteAllocateReplicaShard(RerouteAllocate
...
+                    "Unsupported options provided to REROUTE ALLOCATE " + allocatePrimaryShard.type() + " PRIMARY SHARD: " + properties.keySet());
+            }
+            return new AnalyzedRerouteAllocatePrimaryShard(
+                context.tableInfo,
+                Lists.map(
+                    context.partitionProperties,
+                    p -> p.map(a -> context.exprAnalyzerWithFields.convert(a, context.exprCtx))
+                ),
+                allocatePrimaryShard,
+                acceptDataLoss == null ? Literal.BOOLEAN_FALSE : acceptDataLoss
```
Comment (jeeminso) [inline server/src/main/java/io/crate/analyze/AlterTableRerouteAnalyzer.java:161]: minor: how about
```suggestion
                acceptDataLoss == null ? Literal.BOOLEAN_FALSE : Literal.BOOLEAN_TRUE
```
since we documented that only booleans are valid input.

File: server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        // Allocate the stale primary shard (this is actually happening by the `PROMOTE_REPLICA` command)
+        execute("ALTER TABLE " + tableIdent + " REROUTE PROMOTE REPLICA SHARD 0 ON '" + node + "' WITH (accept_data_loss = true)");
+
+        // wait until the stale shard is allocated and started
+        assertBusy(() -> {
+            execute("SELECT state FROM sys.shards WHERE id = 0 AND primary = true AND state = 'STARTED'");
+            assertThat(response.rowCount()).isEqualTo(1);
+        });
+
+        execute(String.format("SELECT COUNT(*) < %s FROM t", totalDocs));
```
Comment (jeeminso) [inline server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java]: We could also check `count(*) > 0`.

File: docs/cli-tools.rst [hunk 4]
```diff
@@ -212,6 +212,92 @@ Options
...
+.. _cli-crate-shard-synopsis:
+
+Synopsis
+--------
+
+::
+
+   sh$ bin/crate-shard remove-corrupted-data
+   ([--table <FullTableFQN>] (-Ppartitioned_column=value [ , ...]) [--shard-id <ShardId>] | [--dir <TablePath>])
+   [--truncate-clean-translog]
```
Comment (jeeminso) [inline docs/cli-tools.rst]: We are missing documentation for `[--truncate-clean-translog]` and also would be good to test truncating translog.

**Review by jeeminso — APPROVED** [review]
Thank you, left a few comments, otherwise LGTM.

File: server/src/main/java/io/crate/analyze/AlterTableRerouteAnalyzer.java (hunk 2 above)
Comment (seut) [reply server/src/main/java/io/crate/analyze/AlterTableRerouteAnalyzer.java:161]: Wouldn't this be wrong then if the user supplies a `False` value?

File: docs/cli-tools.rst (hunk 4 above)
Comment (seut) [reply docs/cli-tools.rst]: Thanks for catching, done now, see fixup commit.

File: server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java [hunk 5]
```diff
@@ -164,6 +174,10 @@ public Settings onNodeStopped(String nodeName) throws Exception {
 
         logger.info("--> output:\n{}", terminal.getOutput());
 
+        if (truncateTranslog) {
+            assertThat(terminal.getOutput()).contains("Removing existing translog files");
+        }
+
```
Comment (jeeminso) [inline server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java:182]: It would be good to check `else` scenario with `doesNotContain` to make sure translogs are not deleted.

File: server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        String nodeId = clusterService().state().getRoutingNodes().iterator().next().nodeId();
+        String tableIdent = partitioned ? "doc.t PARTITION (p = 1)" : "doc.t";
+        assertThat(terminal.getOutput()).contains("ALTER TABLE " + tableIdent + " REROUTE ALLOCATE STALE PRIMARY SHARD 0 ON '" + nodeId + "' WITH (accept_data_loss = true);");
+
+        // Ensure node has started
+        ensureStableCluster(1, node);
+
+        // Allocate the stale primary shard (this is actually happening by the `PROMOTE_REPLICA` command)
+        execute("ALTER TABLE " + tableIdent + " REROUTE PROMOTE REPLICA SHARD 0 ON '" + node + "' WITH (accept_data_loss = true)");
```
Comment (jeeminso) [inline server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java]: Could you clarify the comment further why we cannot use `ALLOCATE STALE PRIMARY`?

**Review by jeeminso — COMMENTED** [review]
Thank you, left few more comments and also noticed an issue -

executing `REROUTE ALLOCATE STALE PRIMARY` threw an exception
```
UnsupportedFeatureException[Cannot create Plan from AnalyzedStatement "io.crate.analyze.AnalyzedRerouteAllocatePrimaryShard@2cd976ba"  - not supported.]
```

File: app/src/main/dist/bin/crate-shard [hunk 7]
```diff
@@ -0,0 +1,176 @@
+#!/bin/sh
```
Comment (mfussenegger) [inline app/src/main/dist/bin/crate-shard]: Thoughts on adding a `remove-corrupted-shard-data` command to the existing `crate-node` cli to avoid having another bash wrapper script?

Or is there something in the setup that requires a dedicated executable?

File: server/src/main/java/io/crate/analyze/AlterTableRerouteAnalyzer.java [hunk 8]
```diff
@@ -140,6 +141,29 @@ public RerouteAnalyzedStatement visitRerouteAllocateReplicaShard(RerouteAllocate
...
+                                                                         Context context) {
+            AllocatePrimaryShard<Symbol> allocatePrimaryShard =
+                node.map(x -> context.exprAnalyzer.convert((Expression) x, context.exprCtx));
+
+            Map<String, Symbol> properties = allocatePrimaryShard.properties().toMap(HashMap::new);
+            Symbol acceptDataLoss = properties.remove(AllocatePrimaryShard.Properties.ACCEPT_DATA_LOSS);
+            if (!properties.isEmpty()) {
+                throw new IllegalArgumentException(
+                    "Unsupported options provided to REROUTE ALLOCATE " + allocatePrimaryShard.type() + " PRIMARY SHARD: " + properties.keySet());
+            }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/AlterTableRerouteAnalyzer.java]: Just fyi, there's also a  `properties().ensureContainsOnly()`

File: server/src/main/java/io/crate/analyze/AnalyzedRerouteAllocatePrimaryShard.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    public void visitSymbols(Consumer<? super Symbol> consumer) {
+        super.visitSymbols(consumer);
+        consumer.accept(allocatePrimaryShard.shardId());
+        consumer.accept(allocatePrimaryShard.node());
+        consumer.accept(acceptDataLoss);
+        // allocatePrimaryShard.properties() is not visited, because the `acceptDataLoss` value was extracted
+        // from it and other properties are not permitted
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/AnalyzedRerouteAllocatePrimaryShard.java:61]: `ParameterTypes.extract` relies on `visitSymbols` to detect the parameter types of a statement which can be used via pgwire to describe the parameters of a statement.

I think if it doesn't visit the properties something like `ALTER TABLE t REROUTE ALLOCATE EMPTY PRIMARY SHARD 0 ON 'node1' WITH (accept_data_loss = ?)` is broken if a pgwire client asks for the type of `?`

File: server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java (hunk 6 above)
Comment (seut) [reply server/src/test/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommandIT.java]: Changed it, missed to adjust it.

File: server/src/main/java/io/crate/analyze/AnalyzedRerouteAllocatePrimaryShard.java (hunk 9 above)
Comment (seut) [reply server/src/main/java/io/crate/analyze/AnalyzedRerouteAllocatePrimaryShard.java:61]: > I think if it doesn't visit the properties something like ALTER TABLE t REROUTE ALLOCATE EMPTY PRIMARY SHARD 0 ON 'node1' WITH (accept_data_loss = ?) is broken if a pgwire client asks for the type of ?

It is visiting the extracted `accept_data_loss` symbol, which is extracted already from the properties.
I've verified that passing a parameter placeholder works as expected.

File: app/src/main/dist/bin/crate-shard (hunk 7 above)
Comment (seut) [reply app/src/main/dist/bin/crate-shard]: Good point, no there is no need for a dedicated `crate-shard` cli. I'l change it.

File: app/src/assembly/src.xml [hunk 10]
```diff
@@ -41,6 +41,7 @@
         <include>crate</include>
         <include>crate.bat</include>
         <include>crate-node</include>
+        <include>crate-shard</include>
```
Comment (mfussenegger) [inline app/src/assembly/src.xml]: Is this still needed?

File: server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommand.java [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+            } else {
+                throw new ElasticsearchException("Unable to resolve shard id. Wrong folder structure at [ " + path
+                    + " ], expected .../indices/[INDEX-UUID]/[SHARD-ID]");
+            }
+        } else {
+            // otherwise resolve shardPath based on the table name, partition and shard id
+            String indexName = Objects.requireNonNull(tableNameOption.value(options), "Table name is required");
+            List<KeyValuePair> partitionValues = paritionValuesOption.values(options);
+            shardId = Objects.requireNonNull(shardIdOption.value(options), "Shard ID is required");
+            indexMetadata = resolveIndexMetadata(clusterState, indexName, partitionValues);
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommand.java]:
```suggestion
            String fqTableName = Objects.requireNonNull(tableNameOption.value(options), "Table name is required");
            RelationName table = RelationName.fromIndexName(fqTableName);
            List<KeyValuePair> partitionValues = paritionValuesOption.values(options);
            shardId = Objects.requireNonNull(shardIdOption.value(options), "Shard ID is required");
            indexMetadata = resolveIndexMetadata(clusterState, table, partitionValues);
```

Just a suggestion - the `indexName` here confused me a bit and I wasn't sure if it's safe at first given the rename/swap changes.

**Review by mfussenegger — APPROVED** [review]
